File: modules/handlers/dir.py
"""
:project: telegram-onedrive
:author: L-ING
:copyright: (C) 2023 L-ING <hlf01@icloud.com>
:license: MIT, see LICENSE for more details.
"""

from telethon import events
from modules.client import tg_bot
from modules.env import tg_user_name
from modules.utils import check_in_group, check_tg_login, check_od_login, cmd_parser, CMDException
from modules.global_var import dir_res
from modules.onedrive.dir import Dir


@tg_bot.on(events.NewMessage(pattern="/dir", incoming=True, from_users=tg_user_name))
@check_in_group
@check_tg_login
@check_od_login
async def dir_handler(event):
    cmd = cmd_parser(event)
    # /dir
    try:
        if len(cmd) == 1:
            if not Dir.is_temp:
                await event.respond(f'Current directory is `{Dir.path}`')
            else:
                await event.respond(f'Current temporary directory is `{Dir.path}`')
        elif len(cmd) == 2:
            sub_cmd = cmd[1]
            # /dir reset
            if sub_cmd == 'reset':
                Dir.reset()
                await event.respond(f'Directory reset to default `{Dir.path}`')
            # /dir $remote_path
            else:
                remote_path = cmd[1].strip().strip('*')
                if remote_path.startswith('/'):
                    Dir.set_perm_path(remote_path)
                    await event.respond(f'Directory set to `{Dir.path}`')
                else:
                    await event.respond('Directory path should start with /')
        
        elif len(cmd) == 3:
            sub_cmd = cmd[1]
            if sub_cmd == 'temp':
                sub_cmd = cmd[2]
                # /dir temp $remote_path
                if sub_cmd != 'cancel':
                    remote_path = cmd[2].strip().strip('*')
                    if remote_path.startswith('/'):
                        Dir.set_temp_path(remote_path)
                        await event.respond(f'Temporary directory set to `{Dir.path}`')
                    else:
                        await event.respond('Directory path should start with /')
                # /dir temp cancel
                else:
                    Dir.check_temp()
                    await event.respond(f'Directory restored to `{Dir.path}')
            else:
                raise CMDException('Sub command of /dir temp wrong.')
        else:
            raise CMDException('Command /dir format wrong.')
    except CMDException:
        await event.reply(dir_res)

    raise events.StopPropagation

File: src/handlers/dir.rs
/*
:project: telegram-onedrive
:author: L-ING
:copyright: (C) 2024 L-ING <hlf01@icloud.com>
:license: MIT, see LICENSE for more details.
*/

use super::{
    docs::{format_help, format_unknown_command_help},
    utils::{text::cmd_parser, validate_root_path},
};
use crate::{client::OneDriveClient, message::TelegramMessage, state::AppState};
use anyhow::{anyhow, Context, Result};
use grammers_client::InputMessage;
use proc_macros::{check_in_group, check_od_login, check_senders};

pub const PATTERN: &str = "/dir";

#[check_od_login]
#[check_senders]
#[check_in_group]
pub async fn handler(message: TelegramMessage, state: AppState) -> Result<()> {
    let onedrive = &state.onedrive;

    let cmd = cmd_parser(message.text());

    if cmd.len() == 1 {
        // /dir
        show_dir(onedrive, message).await?;
    } else if cmd.len() == 2 {
        if cmd[1] == "reset" {
            // /dir reset
            reset_dir(onedrive, message).await?;
        } else if cmd[1] == "help" {
            // /dir help
            message
                .respond(InputMessage::html(format_help(PATTERN)))
                .await
                .context("help")?;
        } else {
            // dir $root_path
            let root_path = &cmd[1];
            set_dir(onedrive, message, root_path).await?;
        }
    } else if cmd.len() == 3 {
        if cmd[1] == "temp" {
            if cmd[2] == "cancel" {
                // /dir temp cancel
                cancel_temp_dir(onedrive, message).await?;
            } else {
                // /dir temp $path
                let temp_root_path = &cmd[2];
                set_temp_dir(onedrive, message, temp_root_path).await?;
            }
        } else {
            return Err(anyhow!("sub command error")).context(format_unknown_command_help(PATTERN));
        }
    } else {
        return Err(anyhow!("command error")).context(format_unknown_command_help(PATTERN));
    }

    Ok(())
}

async fn show_dir(onedrive: &OneDriveClient, message: TelegramMessage) -> Result<()> {
    let root_path = onedrive.get_root_path(false).await?;
    let is_temp = onedrive.does_temp_root_path_exist().await;

    let response = if is_temp {
        format!("Current directory is {}, and it's temporary.", root_path)
    } else {
        format!("Current directory is {}", root_path)
    };
    message.respond(response.as_str()).await.context(response)?;

    Ok(())
}

async fn reset_dir(onedrive: &OneDriveClient, message: TelegramMessage) -> Result<()> {
    onedrive.reset_root_path().await?;

    let response = format!("Directory reset to default {}", onedrive.default_root_path);
    message.respond(response.as_str()).await.context(response)?;

    Ok(())
}

async fn set_dir(
    onedrive: &OneDriveClient,
    message: TelegramMessage,
    root_path: &str,
) -> Result<()> {
    validate_root_path(root_path).await?;

    onedrive.set_root_path(root_path).await?;

    let response = format!("Directory set to {}", root_path);
    message.respond(response.as_str()).await.context(response)?;

    Ok(())
}

async fn cancel_temp_dir(onedrive: &OneDriveClient, message: TelegramMessage) -> Result<()> {
    onedrive.clear_temp_root_path().await?;

    let response = format!(
        "Temporary directory canceled.\nCurrent directory is {}",
        onedrive.get_root_path(false).await?
    );
    message.respond(response.as_str()).await.context(response)?;

    Ok(())
}

async fn set_temp_dir(
    onedrive: &OneDriveClient,
    message: TelegramMessage,
    temp_root_path: &str,
) -> Result<()> {
    validate_root_path(temp_root_path).await?;

    onedrive.set_temp_root_path(temp_root_path).await?;

    let response = format!("Temporary directory set to {}", temp_root_path);
    message.respond(response.as_str()).await.context(response)?;

    Ok(())
}
